### PR TITLE
perf: reduce native TIA coverage overhead

### PR DESCRIPTION
**Pull request by anmarchenko**
## Summary

- deduplicate source-file work in the native line event hook with a bounded direct cache
- retain 1,024 cached Ruby path objects per test so GC cannot reuse a raw source pointer for a different file
- cache allocation class source resolution across tests
- skip repeated allocation-class bookkeeping before name lookup and hash insertion
- add regression coverage for discarded instruction sequences and an append-only optimization journal

## Performance

Full upstream RuboCop suite with per-test TIA coverage:

- initial (`--warmups 0 --runs 1`): **77.87%** overhead (115.77s baseline / 205.91s instrumented)
- current (`--warmups 1 --runs 4`, medians): **53.46%** overhead (116.85s baseline / 179.31s instrumented)
- overhead reduction: **24.41 percentage points**
- tracer-added time: **90.15s → 62.47s**, saving **27.68s (30.71%)** per full-suite execution
- total instrumented duration: **12.92% lower**

The initial and current results use different sample counts, so these savings are directional rather than a deterministic comparison. The previous implementation reached 48.17% in a one-pair run, but its raw-pointer cache could silently omit files after instruction-sequence collection. The current 53.46% result includes the accepted GC-safe correctness fix.

Every accepted implementation passed Crook's 37-assertion functional gate. Pf2 profiles and every iteration's result paths are retained in the optimization journal.

## Regression guards

- full RuboCop without TIA: 36 assertions passed; comparison inconclusive, not regressed
- Quotes Rails: 36 assertions passed; comparison inconclusive, not regressed

## Validation

- `bundle exec rake compile_ext`
- `bundle exec rspec spec/ddcov/ddcov_spec.rb spec/ddcov/ddcov_with_symlinks_spec.rb` — 38 examples, 0 failures
- focused source-pointer-reuse regression — 2,000/2,000 included files retained
- `bundle exec rake test:main` — 2,308 examples, 0 failures, 1 Linux-only pending
- `ruby-rubocop-coverage` — 37 assertions passed, 53.46% median overhead across four measured pairs

The experiment log and artifact paths are recorded in `docs/performance/ruby-datadog-ci-rubocop-coverage.md`.

### CODE REVIEW

File: ext/datadog_ci_native/datadog_cov.c [hunk 1]
```diff
@@ -185,28 +222,25 @@ static VALUE safely_get_mod_ancestors(VALUE klass) {
...
-
 // This function is called for each class that was instantiated during the test
 // run.
 static int each_instantiated_klass(st_data_t key, st_data_t _value,
                                    st_data_t data) {
   VALUE klass = (VALUE)key;
   struct dd_cov_data *dd_cov_data = (struct dd_cov_data *)data;
 
+  st_data_t cached_files;
+  if (st_lookup(dd_cov_data->klass_files_cache, key, &cached_files)) {
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply ext/datadog_ci_native/datadog_cov.c:262]: dynamically changing the ancestor chain during the test execution is a very rare edge case: we are ready to accept this risk

File: ext/datadog_ci_native/datadog_cov.c [hunk 2]
```diff
@@ -368,12 +439,17 @@ static VALUE dd_cov_stop(VALUE self) {
   // process classes covered by allocation tracing
   st_foreach(dd_cov_data->klasses_table, each_instantiated_klass,
              (st_data_t)dd_cov_data);
+  dd_cov_data->last_allocated_klass = Qnil;
+  memset(dd_cov_data->seen_allocated_klasses, 0,
+         sizeof(dd_cov_data->seen_allocated_klasses));
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply ext/datadog_ci_native/datadog_cov.c:489]: it might make sense to make the classes cache bounded (with a high enough bound), but mutable ancestors are definitely out of scope here

File: ext/datadog_ci_native/datadog_cov.c [hunk 3]
```diff
@@ -145,19 +172,29 @@ static bool record_impacted_file(struct dd_cov_data *dd_cov_data,
...
   if (dd_cov_data->last_filename_ptr == current_filename_ptr) {
     return;
   }
   dd_cov_data->last_filename_ptr = current_filename_ptr;
 
+  size_t cache_index =
+      ((current_filename_ptr >> 4) ^ (current_filename_ptr >> 12)) &
+      (SEEN_FILENAME_CACHE_SIZE - 1);
+  if (dd_cov_data->seen_filename_ptrs[cache_index] == current_filename_ptr) {
+    return;
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply ext/datadog_ci_native/datadog_cov.c:222]: fixed!

File: ext/datadog_ci_native/datadog_cov.c [hunk 4]
```diff
@@ -221,9 +255,22 @@ static int each_instantiated_klass(st_data_t key, st_data_t _value,
...
+    VALUE filename = dd_ci_resolve_const_to_file(klass_name);
+    if (filename == Qnil || !record_impacted_file(dd_cov_data, filename)) {
+      continue;
+    }
+
+    rb_ary_push(files, filename);
   }
 
+  rb_obj_freeze(files);
+  st_insert(dd_cov_data->klass_files_cache, key, (st_data_t)files);
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply ext/datadog_ci_native/datadog_cov.c:307]: it is ok to cache them for the whole time of the program (tests execution time); but memory must be bounded

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@autotest review

File: ext/datadog_ci_native/datadog_cov.c [hunk 5]
```diff
@@ -221,9 +281,28 @@ static int each_instantiated_klass(st_data_t key, st_data_t _value,
...
   }
 
+  rb_obj_freeze(files);
+  // Bound cross-test retention without adding eviction work to cache hits.
+  // Reaching the limit starts a fresh cache generation.
+  if (st_table_size(dd_cov_data->klass_files_cache) >=
+      KLASS_FILES_CACHE_SIZE) {
+    st_clear(dd_cov_data->klass_files_cache);
+  }
+  st_insert(dd_cov_data->klass_files_cache, key, (st_data_t)files);
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply ext/datadog_ci_native/datadog_cov.c:307]: I don't expect Object.const_source_location to temporarily raise, I hope that it is deterministic!

File: ext/datadog_ci_native/datadog_cov.c [hunk 6]
```diff
@@ -185,28 +248,25 @@ static VALUE safely_get_mod_ancestors(VALUE klass) {
...
-
 // This function is called for each class that was instantiated during the test
 // run.
 static int each_instantiated_klass(st_data_t key, st_data_t _value,
                                    st_data_t data) {
   VALUE klass = (VALUE)key;
   struct dd_cov_data *dd_cov_data = (struct dd_cov_data *)data;
 
+  st_data_t cached_files;
+  if (st_lookup(dd_cov_data->klass_files_cache, key, &cached_files)) {
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply ext/datadog_ci_native/datadog_cov.c:262]: I accept this limitation and don't expect it to affect real world test suites

**Comment by anmarchenko** [comment]
@autotest review

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@autotest review

**Review by juan-fernandez — APPROVED** [review]
